### DISCUSSION TITLE
코드 refactoring + NLI, MLM 모델 추가해서 develop branch에 병합.

### DISCUSSION THREAD

**Discussion by intrandom5** [boostcampaitech4nlp2/level1_semantictextsimilarity_nlp-level1-nlp-13#36] [Announcements]
기존 main branch의 코드 정리를 좀 하고, NLI와 MLM 모델을 추가했습니다.
기능은 기존과 똑같이 사용되는 것을 확인했는데, 혹시라도 문제 있으면 해결해서 push 하거나 저에게 알려주면 수정하겠습니다.

### NLI, MLM 모델 추가
추가 사항은 학습 모델로 NLI와 MLM 학습용 모델과 데이터셋을 추가했습니다.
각 모델은 config에서 model_type에 "MLM"과 "BERT_NLI" 옵션을 줘서 학습할 수 있습니다.
MLM 모델 같은 경우에는 EDA를 input으로 받지 않습니다. 설정 상 변경해줘야 하는건 없고 그냥 그렇다고 알려드립니다.

### config 추가 사항
early_stopping과 learning rate shceduler의 파라미터를 일부 config에서 직접 설정할 수 있게 수정.
loss도 config에서 선택할 수 있게 변경.

### EDA 관련.
EDA_output의 저장 결과의 파일명이 "{프로젝트 이름}_e{에포크}_{시간}.csv" 로 저장되도록 변경.
모델에 aux가 input으로 들어가고 있었는데 일단 뺐습니다.
만약 다른 데이터의 feature를 사용하고 싶다면 직접 추가해서 pull request 하는 쪽으로.

### datasets.py 관련.
aux 생성 코드를 함수로 따로 빼서 뒤에 상속 받는 dataset class들이 바로 쓸 수 있게 변경.
stopword 생성 코드를 data_preprocessing.py의 함수를 사용해 짧게 줄임.

### train.py 관련
train.py 코드에서 training 코드가 각 모델 분기 별로 있던 것을 utils의 trainer.py로 빼서 train.py를 조금 더 짧게 줄임.
(그럼에도 전혀 다른 학습인 MLM이 추가되면서 train.py가 아직도 길고 더럽긴 함...)
MLM 학습 같은 경우엔 wandb에 sentence_bert 프로젝트가 아니라 bert-mlm 프로젝트에 결과가 저장되도록 함.(두 모델의 측정 지표가 다르기 때문에 같은 프로젝트를 쓰면 그래프가 더러워짐.)

### inference.py
inference.py가 리뉴얼 되었는데, parser로 config 전달하던 기존 방식에서 현재 우리가 쓰고 있는 yaml 파일을 config로 받아서 수행하도록 변경.
config에서 base_model과 model_type, model_load_path로 inference할 모델을 설정.
config의 test_csv의 경로에 있는 csv 파일을 대상으로 inference를 수행.
그리고 코드가 짧아졌습니다. 결과는 똑같이 output.csv로 저장됩니다.

### config.yaml
마지막으로 현재 최종 코드에 제가 사용하고 있는 config 파일을 공유해 드립니다..
>>>
    test_mode: false

    base_model: klue/roberta-large
    model_type: BERT # SBERT, BERT, BERT_NLI, MLM
    batch_size: 64
    epochs: 20
    early_stopping_patience: 6
    lr_scheduler_patience: 3
    lr_scheduler_factor: 0.2
    lr: 1.0e-5

    loss: MAE # MSE, MAE, BCE, CE

    model_load_path: results/MLM.pt
    model_save_path: results/MLM2STS_from_MLM.pt

    train_csv: NLP_dataset/processed_train.csv
    valid_csv: NLP_dataset/processed_dev.csv
    test_csv: NLP_dataset/processed_test.csv
    stopword: False

    log_name: MLM2STS_from_MLM
    notes: 데이터셋을 MLM으로 사전 훈련 후에 STS 학습 수행.

**Comment by papari1123** [comment]
리팩토링이 참 귀찮은 작업인데, 고생했고 오전 실험 끝나고 PULL해서 테스트해볼게~
